scripts/lintcommit.lua

```diff
@@ -0,0 +1,145 @@
+local function commit_message_is_ok(commit_message)
+  local commit_split = vim.split(commit_message, ":")
+
+  -- Return true if the type is vim-patch since most of the normal rules don't
+  -- apply.
+  if commit_split[1] == "vim-patch" then
+    return true
+  end
+
+  -- Check that message isn't too long.
+  if commit_message:len() > 80 then
+    vim.cmd([[verbose echo "Commit message is too long, a maximum of 80 characters is allowed. Commit message:"]])
+    return false
+  end
+
+  -- Return false if no colons are detected.
+  if vim.tbl_count(commit_split) < 2 then
+    vim.cmd([[verbose echo "Commit message does not include colons. Commit message:"]])
+    return false
+  end
+
+  local before_colon = commit_split[1]
+  local after_colon = commit_split[2]
+
+  -- Check if commit introduces a breaking change.
+  if vim.endswith(before_colon, "!") then
+    before_colon = before_colon:sub(1, -2)
+  end
+
+  -- Check if type is correct
+  local type = vim.split(before_colon, "%(")[1]
+  local allowed_types = {"build", "ci", "docs", "feat", "fix", "perf", "refactor", "revert", "test", "chore"}
+  if not vim.tbl_contains(allowed_types, type) then
+    vim.cmd([[verbose echo "Commit type is not recognized. Allowed types are: build, ci, docs, feat, fix, perf, refactor, revert, test, chore. Commit message:"]])
+    return false
+  end
+
+  -- Check if scope is empty
+  if before_colon:match("%(") then
+    local scope = vim.trim(before_colon:match("%((.*)%)"))
+    if scope == '' then
+      vim.cmd([[verbose echo "Scope can't be empty. Commit message:"]])
+      return false
+    end
+  end
+
+  -- Check that description doesn't end with a period
+  if vim.endswith(after_colon, ".") then
+      vim.cmd([[verbose echo "Description ends with a period (\".\"). Commit message:"]])
+    return false
+  end
+
+  -- Check that description has exactly one whitespace after colon, followed by
+  -- a lowercase letter and then any number of letters.
+  if not string.match(after_colon, '^ %l%a*') then
+      vim.cmd([[verbose echo "There should be one whitespace after the colon and the first letter should lowercase. Commit message:"]])
+    return false
+  end
+
+  return true
+end
+
+local function main()
+  local current_branch = vim.trim(vim.fn.system({'git', 'branch', '--show-current'}))
+  local ancestor_commit = vim.trim(vim.fn.system({'git', 'merge-base', 'origin/master', current_branch}))
+  local commits_str = vim.trim(vim.fn.system({'git', 'rev-list', ancestor_commit .. ".." .. current_branch}))
+
+  local commits = {}
+  for substring in commits_str:gmatch("%S+") do
+     table.insert(commits, substring)
+  end
+
+  for _, commit_hash in ipairs(commits) do
+    local message = vim.trim(vim.fn.system({'git', 'show', '-s', '--format=%s' , commit_hash}))
```

Comment:
need to check for errors here.

```
  for _, commit_hash in ipairs(commits) do
    local message = vim.trim(vim.fn.system({'git', 'show', '-s', '--format=%s' , commit_hash}))
    if vim.v.shell_error ~= 0 then
      vim.api.nvim_echo({{'invalid commit-id: '..commit_hash,''}},false,{})
    elseif not commit_message_is_ok(message) then
      vim.api.nvim_echo({{message,''}},false,{})
      vim.cmd("cquit 1")
    end
  end
```

Reply:
I've added the error check, but I couldn't get `vim.api.nvim_echo` to work. Reverted to verbose echo.